### PR TITLE
[WIP] Microkernels for CUDA target

### PR DESCRIPTION
**Pull request by grypp**
This PR introduces microkernels for CUDA target. It is a very first step, and there are many TODOs, but it works e2e `linalg.matmul`.

It implements followings:
- Introduce a new Op `MicroKernelOp` in `IREECodegen` dialect.
- Adds new compilation pipeline `LLVMGPU_Microkernel`, and sets it if micro kernel flag is enabled.
- Lowers `linalg.matmul` into `MicroKernelOp` after bufferization.
- Lowers `MicroKernelOp` into a trampoline function call before LLVM/NVVM generation.
- Inserts pre-compiled PTX microkernels just before ptx compilation.

PR also introduces CUDA microkernels code. The idea is to call optimized CUDA codes here.

PR inserts microkernels as PTX. I confirmed that `ptxas` inlines trampoline function and the microkernels when I look at the output sass code.

### CODE REVIEW

**Comment by benvanik** [comment]
Can't comment much on the microkernel op stuff, but we should have a chat about this so we can dig into some of the goo layers. 

TLDR is for linking we need to align this with the infrastructure in place for custom linking that we use for either full dispatch replacement on CUDA or partial substitution on the CPU. See https://github.com/iree-org/iree/tree/main/samples/custom_dispatch.

The only work that should be required is extending CUDATarget.cpp to handle multiple IREE::HAL::ExecutableObjectAttrs like the LLVMCPUTarget.cpp already does - if it's as simple as concatenating the PTX blobs that should be really easy.

Steps:
* the microkernel needs to get turned into a call (https://github.com/iree-org/iree/blob/main/samples/custom_dispatch/cpu/embedded/example_hal.mlir#L187)
* the external function needs to be declared (https://github.com/iree-org/iree/blob/f19cc5b977e5210411068c820dab688cec5cc377/samples/custom_dispatch/cpu/embedded/example_hal.mlir#L152-L156)
* the object files to link in need to be attached (https://github.com/iree-org/iree/blob/f19cc5b977e5210411068c820dab688cec5cc377/samples/custom_dispatch/cpu/embedded/example_hal.mlir#L61-L74)

This then supports multiple external objects coming in for if multiple microkernels are used in the same dispatch. It'll also let us link things in once when we are combining PTX blobs (which we really need to do...).

Actually producing the ptx blobs for the microkernels is going to be the difficult part and will need some build infrastructure. The custom_dispatch sample shells out to nvcc but that doesn't scale beyond a few - if we can keep this bounded to just 5-10 microkernels then we should be ok keeping it simple, and if it extends beyond that this should probably live in a separate location/layer of the stack (we don't want to rebuild cublas :). The biggest issue this entire approach has is that we want to be compiling the microkernels with the same target configuration being used by the compiler and we can't enumerate all possible target combinations. How we organize the files, decide what to compile with what settings, and how to select from those is going to be the biggest chunk of work. Again if we can keep this set small then things stay simple - otherwise this will need to live separate (some big cmake flag to enable as opt-in, etc).

The source files themselves shouldn't be in runtime/ as they aren't related to runtime at all. `compiler/Codegen/Microkernels/CUDA/` is probably a good place to put them for now.

**Comment by benvanik** [comment]
The other thing worth investigating here is bitcode - if we could get in before PTX that'd be much better and let us build our own libdevice.bc-equivalent. Getting nvcc to spit out LLVM bitcode instead of PTX would mean that we'd be able to perform linking ourselves via LLVM and avoid the need for one file per microkernel and at least somewhat control the target configuration (lots still baked in to the bitcode but much less than PTX). The object file reference attributes above would just change to .bc instead of .ptx and we'd link them just like how we link libdevice: https://github.com/google/iree/blob/21dd0aae7fd4914873ad5caa75ab10bbc40a0a09/compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp#L102-L128

**Comment by stellaraccident** [comment]
> Actually producing the ptx blobs for the microkernels is going to be the difficult part and will need some build infrastructure.

What Ben says: the code organization so everything plays together is the hard (and annoying) part and things break down a lot if cutting against the grain.

This is cool work, and I'm willing to help on the build esoterica as it comes up. It's non trivial and requires a bit of thought, but mirrors a lot of what I helped put in place for the CPU targets. I'd recommend getting the actual code and layering aligned to Ben's prescription and doing the simplest possible thing on the build side (guarded by an experimental flag) and then passing off to one of us who specializes in making that stuff more tractable.

**Comment by grypp** [comment]
Thanks a lot for your prompt review. This is really a spike implementation that I implemented in a couple of days. It is great it started a discussion here. I would be happy if we meet up to chat about this. Thanks for showing me the dispatch infra, it looks cool! It is really useful to have multiple ptx for different targets.

Let me explain the rationale behind choosing PTX. In fact, I also prefer the bitcode :) The idea of having our own libdevice fits perfectly with our toolchain. The bitcode can be linked by llvm and the microkernels can be verified in the compile time and many more benefits. Bitcode gives us full control.

But the only reason for this project is performance, so I chose nvcc whose performance is guaranteed, unfortunately, it can only generate PTX. There is clang, it generates bitcode but I don't know whether it generates code as well as nvcc, I haven't seen any metric. nvcc is based on nvidia's proprietary llvm which is different from open source llvm what we have. Also, I don't know if clang supports the latest and greatest CUDA and GPU features :) The hopper GPU is already here, its ISA is large. So the PTX route made more sense since the purpose of this project was to use the latest CUDA codes as microkernels. 

Just to give an idea about the clang vs nvcc compilation, I put their flow. We have only clang to generate bitcode that I wanted to show. The public nvcc does not generate a bitcode. Here, `ptxas` is nvidia ptx compiler.
```
input.cu --[clang]--> input.ll --[llvm]--> input.ptx  --[ptxas]--> binary.sass
```
```
input.cu --[nvcc]--> input.ptx --[ptxas]--> binary.sass
```

Then let me continue with this project like that: I can do benchmarks with nvcc and clang on certain CUDA codes like cutlass. If clang performs well, then we can use bitcode.

**Comment by benvanik** [comment]
Experiments are fine, but the situation in which we require specific proprietary nvidia compilers as part of our core compilation flow is not likely to go much beyond experimental in IREE's core codebase. Having a black-box compiler in our low-level infrastructure is just like if clang passed IR to LLVM which then under the covers shells out to gcc to do codegen for specific functions - it may work, but it's not sustainable :) Nesting compiler flows like that is fine for wrapping layers to do (a user wants to compile half their code with gcc and half with clang then passes all the .o's to ld) but not really something we want in the core compiler itself.

The bitcode approach meets more of the needs we have as a compiler and will be what we want to focus real development on for sustained usage. If there are performance deltas between the proprietary compiler and LLVM those should be raised with LLVM and used to improve the compiler. But we should support both (and the linking mechanism can) but leave it to users/libraries/frameworks/etc to make the more complex decisions around compiler mixes, timelines, tradeoffs in configuration and control, etc rather than us imposing that upon all users. As we build out more extension mechanisms (like microkernels/linking) we can use these experiments as driving examples for user workflows to ensure we have the right points for hooking in. We just don't want to be the ones maintaining and supporting those :)

The custom_dispatch sample shows how to use entire PTX blobs with no changes needed - that's still the best route for "I just want to use nvcc to compile a kernel" - microkernels and mixing compilers in the same local function unit is always going to be sketchy. You can see if you can do some of your experiments using that and KISS :)

**Comment by stellaraccident** [comment]
> Thanks a lot for your prompt review. This is really a spike implementation that I implemented in a couple of days. It is great it started a discussion here. I would be happy if we meet up to chat about this. Thanks for showing me the dispatch infra, it looks cool! It is really useful to have multiple ptx for different targets.
> 
> Let me explain the rationale behind choosing PTX. In fact, I also prefer the bitcode :) The idea of having our own libdevice fits perfectly with our toolchain. The bitcode can be linked by llvm and the microkernels can be verified in the compile time and many more benefits. Bitcode gives us full control.
> 
> But the only reason for this project is performance, so I chose nvcc whose performance is guaranteed, unfortunately, it can only generate PTX. There is clang, it generates bitcode but I don't know whether it generates code as well as nvcc, I haven't seen any metric. nvcc is based on nvidia's proprietary llvm which is different from open source llvm what we have. Also, I don't know if clang supports the latest and greatest CUDA and GPU features :) The hopper GPU is already here, its ISA is large. So the PTX route made more sense since the purpose of this project was to use the latest CUDA codes as microkernels.
> 
> Just to give an idea about the clang vs nvcc compilation, I put their flow. We have only clang to generate bitcode that I wanted to show. The public nvcc does not generate a bitcode. Here, `ptxas` is nvidia ptx compiler.
> 
> ```
> input.cu --[clang]--> input.ll --[llvm]--> input.ptx  --[ptxas]--> binary.sass
> ```
> 
> ```
> input.cu --[nvcc]--> input.ptx --[ptxas]--> binary.sass
> ```
> 
> Then let me continue with this project like that: I can do benchmarks with nvcc and clang on certain CUDA codes like cutlass. If clang performs well, then we can use bitcode.

Let's review and look at results/options before committing to a path. These things have long range costs and implications, and I'd like to make the decision with facts.

**Comment by grypp** [comment]
I updated my PR, here are the changes:
- Created a bitcode library, used `clang++` to compile CUDA microkernels
- Linked iree generated code to the bitcode library (no more inline PTX)
- Moved CUDA microkernels to `runtime/src/builtins/uGPUkernels`

I still need to compare performance of nvcc vs clang++ to see if this is the best choice.

@benvanik do you think do I still need `#hal.executable.object` here?

**Comment by MaheshRavishankar** [comment]
> I updated my PR, here are the changes:
> 
> * Created a bitcode library, used `clang++` to compile CUDA microkernels
> * Linked iree generated code to the bitcode library (no more inline PTX)
> * Moved CUDA microkernels to `runtime/src/builtins/uGPUkernels`
> 
> I still need to compare performance of nvcc vs clang++ to see if this is the best choice.
> 
> @benvanik do you think do I still need `#hal.executable.object` here?

Some follow up. Guray and I chatted about this. I am working on the path to call microkernels from CPU backend. Its very similar to what Guray has, just a bit more flushed out. Doesnt solve a lot of the issues w.r.t this example though. Agree with Ben, this is probably not want to use microkernels for, but I do know you want to use for something else, and this was a way for you to put in the scaffolding needed to experiment. I'd suggest instead of trying to hook this into the compiler directly, just write an `.cu` file and experiment with performance and microkernels you want there. We can then look into what you want the generated code to look like and how to get the compiler to hit that... Doing this in a compiler directly is harder, so would be one way to justify the investment.

**Comment by benvanik** [comment]
Please include me in any future meetings too - I've got a lot of state on these things and don't think anyone particularly enjoys me writing up big responses after the fact :)

Agreed with Mahesh that we still want to better quantify what this route would entail as all the complex parts come when it scales beyond a single reference kernel or single target. As recommended before a way to side-step the issues for now is using the object linking support to reference external files as that can be done without needing to solve the bigger issues until ready (just point it at files on disk or embed the file as a binary blob inside of the IR from an earlier stage). We are going to end up with one way to do this kind of thing such that users and bring-your-own-codegen approaches can be done in a way consistent with what the compiler does - and that means the compiler _also_ needs to be consistent with how it does things wherever possible (meaning if we have a story for linking custom microkernels on the CPU the CUDA side needs to follow that).

On the CPU side we've already done some work to author libraries to bitcode that are retargetable and unsurprisingly it's nearly identical to what CUDA itself does for libdevice. We will need to follow that pattern here. The two bitcode files you have in this PR, for example, differ by exactly 5 of 7652 bytes (something we see in a lot of our codegen output, too) and if you crack them open you'll see those 5 bytes are just the target-cpu and target-features strings - otherwise the entire rest of each file is identical. It's easy to ignore 8KB for now but if each new microkernel added for each new platform is 8KB then you can imagine if we had even 20 microkernels and were targeting 8 concurrent target configurations things get out of hand fast (we don't want MB of churn in the .git every time we tweak a microkernel or integrate and have a new version of LLVM that generates ever so slightly different byte patterns). In other words, we should be looking to generate a single LLVM module that is parameterized on compiler-fed values instead of stamping out an entire copy of each library for each target configuration. This is a good example of the kind of thing we need to understand more before building out infrastructure.

It's informative to take a look at how libdevice handles this - if you crack it open (llvm-dis on libdevice.10.bc) you'll see normal control flow based on placeholders populated by later stages of the compiler:
```llvm
@.str.1 = private unnamed_addr constant [12 x i8] c"__CUDA_ARCH\00", align 1
define i32 @__nv_isinfd(double %a) #0 {
  %1 = call i32 @__nvvm_reflect(ptr @.str.1)
  %2 = icmp eq i32 %1, 200
  br i1 %2, label %9, label %3
```

For more information, see:
https://llvm.org/docs/NVPTXUsage.html#linking-with-libdevice
specifically __nvvm_reflect, which matches what we do on the CPU side during linking:
https://github.com/iree-org/iree/tree/main/runtime/src/iree/builtins/device#link-time-configuration
(as is a common saying around here: GPU and CPU have way more in common in a lot of aspects than people with experience in either side tend to think :)

This means that we shouldn't use preprocessor macros to switch code behavior but instead normal if statements conditional on a placeholder - for a concrete reason why there's ~250 such queries as above in libdevice saving them 250 copies of libdevice (400KB instead of 100MB) while having no impact on runtime performance as all of the branches are evaluated at compile-time and dead paths are stripped. There may be some very coarse reasons to fork the files - like 32-bit vs 64-bit - but it needs to be an exceedingly good reason. We can have our own __nvvm_reflect-alike that we can populate as we want. This also intersects with executable-level constants (aka specialization constants) and other runtime-based selection, and that too is solved in a way similar to the __nvvm_reflect just done at runtime instead of in the offline linker: https://github.com/google/iree/blob/b07528bfac7cc72362c82bdc7c769dec033690c0/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92. The benefits of this kind of stuff will make more sense if we are using JIT LTO to specialize things at load-time: https://developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library/ - I'm likely going to enable on the runtime side so that we can do full load-time specialization soon and that means that microkernels will be able to be conditionally selected/specialized all the way at runtime if desired so long as we wire it up right.

Incidentally, the same thing we need to do for that in the hand-authored code is what we should be doing in codegen: generate the potential paths and then prune when appropriate and parameterize much more. The more we make both the hand-authored side and the compiler-generated side the same the easier things are to maintain and scale. And it's strictly less work :)

As an aside because it can be tricky to pick apart the whys just from the dir tree, but `compiler/Codegen/Microkernels/CUDA/` is where these belong - the builtins/ in the runtime are there because they are shared with the runtime (used by VMVX, etc) - these would only exist in the compiler and have no relation to the runtime. If you find yourself wanting to put things in the runtime that's a big deal so as a general rule it's best to avoid that :)

But all that said, experimenting in a lightweight way and building out the experience of how to piece together the puzzle should be the priority - we need things like the number of microkernels we estimate to exist, how they are produced, who produces them (end users, library authors, us, vendors, etc), what repo they live in (if this grows in scope to something like aitemplate then it's an entire project, vs if it's model architecture specific and belongs in some model library, vs it really is just for conv/matmul), the variance of the final compiled microkernels across different targets, etc etc, and I think what you have here as an experiment would let you do some of that before we lock into any particular path. We _also_ need to solve this for the CPU side for all the same reasons and it's why we haven't fully finished that: it's easy to wire up flags and function calls but much more difficult to decide what those function calls should be :)

What we'll want to do is cut what lands down to the bare minimum to allow for user-specified objects to be linked in to our generated code using the same mechanism we have on the CPU side. We'll then want to experiment with what we link in. Then we can work through the goo and get things wired up for default paths. E.g., instead of trying to productionize these bitcode files right now add a flag that points at a file and adds the `#hal.executable.object` attrs - then you can still iterate on all the rest of the stuff and not complicate builds until we know how to mitigate some of the complexity.

**Comment by MaheshRavishankar** [comment]
> It's informative to take a look at how libdevice handles this - if you crack it open (llvm-dis on libdevice.10.bc) you'll see normal control flow based on placeholders populated by later stages of the compiler:
> 
> ```llvm
> @.str.1 = private unnamed_addr constant [12 x i8] c"__CUDA_ARCH\00", align 1
> define i32 @__nv_isinfd(double %a) #0 {
>   %1 = call i32 @__nvvm_reflect(ptr @.str.1)
>   %2 = icmp eq i32 %1, 200
>   br i1 %2, label %9, label %3
> ```
> 
> For more information, see: [llvm.org/docs/NVPTXUsage.html#linking-with-libdevice](https://llvm.org/docs/NVPTXUsage.html#linking-with-libdevice) specifically __nvvm_reflect, which matches what we do on the CPU side during linking: [`main`/runtime/src/iree/builtins/device#link-time-configuration](https://github.com/iree-org/iree/tree/main/runtime/src/iree/builtins/device?rgh-link-date=2023-02-07T04%3A44%3A08Z#link-time-configuration) (as is a common saying around here: GPU and CPU have way more in common in a lot of aspects than people with experience in either side tend to think :)
> 
> This means that we shouldn't use preprocessor macros to switch code behavior but instead normal if statements conditional on a placeholder - for a concrete reason why there's ~250 such queries as above in libdevice saving them 250 copies of libdevice (400KB instead of 100MB) while having no impact on runtime performance as all of the branches are evaluated at compile-time and dead paths are stripped. There may be some very coarse reasons to fork the files - like 32-bit vs 64-bit - but it needs to be an exceedingly good reason. We can have our own __nvvm_reflect-alike that we can populate as we want. This also intersects with executable-level constants (aka specialization constants) and other runtime-based selection, and that too is solved in a way similar to the __nvvm_reflect just done at runtime instead of in the offline linker: [google/iree@`b07528b`/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92](https://github.com/google/iree/blob/b07528bfac7cc72362c82bdc7c769dec033690c0/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92). The benefits of this kind of stuff will make more sense if we are using JIT LTO to specialize things at load-time: [developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library](https://developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library/) - I'm likely going to enable on the runtime side so that we can do full load-time specialization soon and that means that microkernels will be able to be conditionally selected/specialized all the way at runtime if desired so long as we wire it up right.
> 

Ha! Almost verbatim what I told Guray in our 1:1!

**Comment by stellaraccident** [comment]
> > It's informative to take a look at how libdevice handles this - if you crack it open (llvm-dis on libdevice.10.bc) you'll see normal control flow based on placeholders populated by later stages of the compiler:
> > ```llvm
> > @.str.1 = private unnamed_addr constant [12 x i8] c"__CUDA_ARCH\00", align 1
> > define i32 @__nv_isinfd(double %a) #0 {
> >   %1 = call i32 @__nvvm_reflect(ptr @.str.1)
> >   %2 = icmp eq i32 %1, 200
> >   br i1 %2, label %9, label %3
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For more information, see: [llvm.org/docs/NVPTXUsage.html#linking-with-libdevice](https://llvm.org/docs/NVPTXUsage.html#linking-with-libdevice) specifically __nvvm_reflect, which matches what we do on the CPU side during linking: [`main`/runtime/src/iree/builtins/device#link-time-configuration](https://github.com/iree-org/iree/tree/main/runtime/src/iree/builtins/device?rgh-link-date=2023-02-07T04%3A44%3A08Z#link-time-configuration) (as is a common saying around here: GPU and CPU have way more in common in a lot of aspects than people with experience in either side tend to think :)
> > This means that we shouldn't use preprocessor macros to switch code behavior but instead normal if statements conditional on a placeholder - for a concrete reason why there's ~250 such queries as above in libdevice saving them 250 copies of libdevice (400KB instead of 100MB) while having no impact on runtime performance as all of the branches are evaluated at compile-time and dead paths are stripped. There may be some very coarse reasons to fork the files - like 32-bit vs 64-bit - but it needs to be an exceedingly good reason. We can have our own __nvvm_reflect-alike that we can populate as we want. This also intersects with executable-level constants (aka specialization constants) and other runtime-based selection, and that too is solved in a way similar to the __nvvm_reflect just done at runtime instead of in the offline linker: [google/iree@`b07528b`/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92](https://github.com/google/iree/blob/b07528bfac7cc72362c82bdc7c769dec033690c0/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92). The benefits of this kind of stuff will make more sense if we are using JIT LTO to specialize things at load-time: [developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library](https://developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library/) - I'm likely going to enable on the runtime side so that we can do full load-time specialization soon and that means that microkernels will be able to be conditionally selected/specialized all the way at runtime if desired so long as we wire it up right.
> 
> Ha! Almost verbatim what I told Guray in our 1:1!

@ThomasRaoux There's a lot of what Ben and Mahesh are saying that matches what I was referring to with you today about the way to get to a concrete design on this stuff. I think that some lighter weight prototyping to arrive at the minimal microkernel surface and interface is probably the more interesting thing (and can be prototyped in C) vs the plumbing to make it all work in one step.

**Comment by ThomasRaoux** [comment]
> > > It's informative to take a look at how libdevice handles this - if you crack it open (llvm-dis on libdevice.10.bc) you'll see normal control flow based on placeholders populated by later stages of the compiler:
> > > ```llvm
> > > @.str.1 = private unnamed_addr constant [12 x i8] c"__CUDA_ARCH\00", align 1
> > > define i32 @__nv_isinfd(double %a) #0 {
> > >   %1 = call i32 @__nvvm_reflect(ptr @.str.1)
> > >   %2 = icmp eq i32 %1, 200
> > >   br i1 %2, label %9, label %3
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > For more information, see: [llvm.org/docs/NVPTXUsage.html#linking-with-libdevice](https://llvm.org/docs/NVPTXUsage.html#linking-with-libdevice) specifically __nvvm_reflect, which matches what we do on the CPU side during linking: [`main`/runtime/src/iree/builtins/device#link-time-configuration](https://github.com/iree-org/iree/tree/main/runtime/src/iree/builtins/device?rgh-link-date=2023-02-07T04%3A44%3A08Z#link-time-configuration) (as is a common saying around here: GPU and CPU have way more in common in a lot of aspects than people with experience in either side tend to think :)
> > > This means that we shouldn't use preprocessor macros to switch code behavior but instead normal if statements conditional on a placeholder - for a concrete reason why there's ~250 such queries as above in libdevice saving them 250 copies of libdevice (400KB instead of 100MB) while having no impact on runtime performance as all of the branches are evaluated at compile-time and dead paths are stripped. There may be some very coarse reasons to fork the files - like 32-bit vs 64-bit - but it needs to be an exceedingly good reason. We can have our own __nvvm_reflect-alike that we can populate as we want. This also intersects with executable-level constants (aka specialization constants) and other runtime-based selection, and that too is solved in a way similar to the __nvvm_reflect just done at runtime instead of in the offline linker: [google/iree@`b07528b`/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92](https://github.com/google/iree/blob/b07528bfac7cc72362c82bdc7c769dec033690c0/compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir/#L65-L92). The benefits of this kind of stuff will make more sense if we are using JIT LTO to specialize things at load-time: [developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library](https://developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library/) - I'm likely going to enable on the runtime side so that we can do full load-time specialization soon and that means that microkernels will be able to be conditionally selected/specialized all the way at runtime if desired so long as we wire it up right.
> > 
> > 
> > Ha! Almost verbatim what I told Guray in our 1:1!
> 
> @ThomasRaoux There's a lot of what Ben and Mahesh are saying that matches what I was referring to with you today about the way to get to a concrete design on this stuff. I think that some lighter weight prototyping to arrive at the minimal microkernel surface and interface is probably the more interesting thing (and can be prototyped in C) vs the plumbing to make it all work in one step.

That makes sense. I'll talk to Guray tomorrow and we'll figure out a lighter version of the prototype.

For the linking part, note that even with ptx we can wrap it in an llvm bc file and do linking the same way we handle libdevice. How we generate the code is a separate problem and out of scope for the first prototype.

**Comment by MaheshRavishankar** [comment]
@grypp might be useful for you to track https://github.com/iree-org/iree/pull/12026 . That bsaically does a lot of what you have here. For now just track it, once I land that you can rebase on top of that PR.

**Comment by benvanik** [comment]
Related to this, I've been experimenting with building clang as part of our compiler build - it's an extra ~3mins on my machine, but if people are going to start going hog wild with microkernels it may be the only option. We'd build clang ourselves (like we do lld), invoke that as part of cmake with a very constrained environment, have it produce the bitcode files we need for use by the compiler, and then not have to check in bitcode. It'd not change any of the above with respect to what kind of code we put in the thing we're compiling (still needs to be parameterized for compile/link/runtime-time so we don't invoke clang 10000x times), but would solve the .git pain for both CPU and GPU. I'll keep fiddling with it.

**Comment by grypp** [comment]
@benvanik using the `__nvvm_reflect` type makes sense, @MaheshRavishankar  mentioned exactly the same thing. I will implement a prototype for this. As sm_80 is server gpu I did not consider MB of the bitcode. But why not reduce size/number of bitcode if it is possible, I am going to include this in my prototype. Thanks! 
Nevertheless, I envision the microkernel code of sm_80 vs sm_90 will be significantly different.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I understand this is a draft, but would be really good to use a uniform approach for all micro kernel hand off. Lets chat about this offline.

**Comment by ThomasRaoux** [comment]
> I understand this is a draft, but would be really good to use a uniform approach for all micro kernel hand off. Lets chat about this offline.

Have all the new ops landed? It wasn't the case when this started and the point of the prototype is to figure out the pieces specific to GPU. It would definitely move to using the common ops to land.

**Comment by MaheshRavishankar** [comment]
Yup. All the ops have landed and I am connecting them to vmvx e2e now. There still might be some changes to the specifics of the op. So this has to tag along behind the micro kernel work already in flight before it can land. Also I think the explicit recommendation was to not try to do this in the compiler first, but rather do this in just a CUDA kernel outside IREE. You should be able to write/evaluate micro kernels completely independent of IREE

**Comment by ThomasRaoux** [comment]
> Yup. All the ops have landed and I am connecting them to vmvx e2e now. There still might be some changes to the specifics of the op. 

great, perfect timing.

> So this has to tag along behind the micro kernel work already in flight before it can land. 

yep

> Also I think the explicit recommendation was to not try to do this in the compiler first, but rather do this in just a CUDA kernel outside IREE. You should be able to write/evaluate micro kernels completely independent of IREE

I don't want to speak for Guray but he has been mixing pure cuda and compiler for experiments. I think giving him the freedom to experiment is important so I don't think we need to change the course of actions at this point.

**Comment by benvanik** [comment]
great if there's progress being made - experiments are awesome and we need more of them - they just don't need to land in main :)

**Comment by grypp** [comment]
It's still in draft :) I just did mid check-in to point out my changes. I need to resolve a few question marks before landing this into the main. 

TODOs
1) Integrate with the cutlass. Currently I can call for one tiles sizes, but it needs to be generic. Big thanks to @manishucsd who is helping me to get there.
1) Calculate the size of shared memory as cutlass in the host code.
2) Fuse `linalg.fill` + `linalg.matmul` in the microkernel.
3) Find a way for `linalg.matmul` consumers to use shared memory filled by the microkernel. This will be proof of fusing `linalg.matmul` + consumers.
4) Use @MaheshRavishankar microkernel Ops.
5) #12222 which looks really cool. I need it to link the microkernels. 
6) Benchmark PTX nvcc or vs bitcode clang++. 

Once I'm done with these items, I'd like to show the performance.

If everyone thinks that having microkernel will improve performance quickly, we can think of what we can do to land it.

**Comment by benvanik** [comment]
There's been a lot of discussion here so probably worth summarizing:
Experimenting with microkernels is interesting, infrastructure for microkernels is interesting, user-serviceable features for making microkernels possible to integrate end-to-end without lots of code changes is interesting, evaluating what cutlass is doing and how its techniques may help us author our own microkernels sustainably or improve our codegen is interesting, but compiling cutlass kernels to ptx and then splicing them into codegen results as part of the core compiler flow is not a production path we'll support. Doesn't mean users can't and libraries layered on top of IREE can't but it's not a piece of engineering we want in the compiler.

**Comment by ThomasRaoux** [comment]
> There's been a lot of discussion here so probably worth summarizing: Experimenting with microkernels is interesting, infrastructure for microkernels is interesting, user-serviceable features for making microkernels possible to integrate end-to-end without lots of code changes is interesting, evaluating what cutlass is doing and how its techniques may help us author our own microkernels sustainably or improve our codegen is interesting, but compiling cutlass kernels to ptx and then splicing them into codegen results as part of the core compiler flow is not a production path we'll support. Doesn't mean users can't and libraries layered on top of IREE can't but it's not a piece of engineering we want in the compiler.

I synced with Ben offline. In summary we agreed that we don't want to bake cutlass based group level function (micro-kernels may be a confusing term) into IREE as that would add dependency between IREE and cutlass that we don't want. However we *can* and will want to use cutlass based group level functions by explicitly replacing linalg.matmul by a micro-kernel kind of op referencing to a cutlass based kernel that would include extra metadata like the tile size. This will allow to layer cutlass usage on top of iree and provides a more modular way to add such dependencies.

Overall the steps done by @grypp are still relevant and I think the order of tasks roughly match this. The integration will force us to get the general micro-kernel infrastructure working and test the layering. One open question will be where the layer to decides on using those micro-kernels will live but this can be deferred. 

@benvanik feel free to correct/add to the summary.

**Comment by vinodgro** [comment]
@ThomasRaoux 
> However we can and will want to use cutlass based group level functions by explicitly replacing linalg.matmul by a micro-kernel kind of op referencing to a cutlass based kernel that would include extra metadata like the tile size. 

so the proposal is to map linalg.matmul style ops with  a library call that is internally based on cutlass instantiations?

**Comment by ThomasRaoux** [comment]
> @ThomasRaoux
> 
> > However we can and will want to use cutlass based group level functions by explicitly replacing linalg.matmul by a micro-kernel kind of op referencing to a cutlass based kernel that would include extra metadata like the tile size.
> 
> so the proposal is to map linalg.matmul style ops with a library call that is internally based on cutlass instantiations?

Correct, the library call would be a just wrapper around an instantiation of threadblock level cutlass function.

We would pre-compile (to ptx or llvm ir) some of those functions for a few sets of tile sizes/pipeline depth. Then those can replace linalg.matmul op after tiling to block level. This allows us to use cutlass while doing codegen for the rest of the kernel which may include anything we want to fuse.

**Comment by benvanik** [comment]
Closing stale PR.
